node/libs/storage/src/block_store/mod.rs

```diff
@@ -188,19 +239,34 @@ impl BlockStore {
         &self,
         ctx: &ctx::Ctx,
         block: validator::FinalBlock,
-    ) -> ctx::OrCanceled<()> {
+    ) -> ctx::Result<()> {
         let number = block.header().number;
-        sync::wait_for(ctx, &mut self.subscribe(), |queued_state| {
-            queued_state.next() >= number
-        })
-        .await?;
+        {
+            let sub = &mut self.subscribe();
+            let queued_state =
+                sync::wait_for(ctx, sub, |queued_state| queued_state.next() >= number).await?;
+            if queued_state.next() != number {
+                return Ok(());
+            }
+            if queued_state.last.as_ref().map(|qc| qc.header().hash()) != block.header().parent {
+                return Err(
+                    anyhow::format_err!("block.parent doesn't match the previous block").into(),
+                );
+            }
+            // Verify the message without verifying the signatures, to avoid doing it twice.
+            // TODO(gprusak): consider moving the signature verification here for resiliency.
+            block
+                .justification
+                .message
+                .verify(&self.genesis, /*allow_past_forks=*/ true)?;
```

Comment:
I've changed my mind, do double check the block signature here. Defense in depth and all that.

Reply:
done